File: prep/implementation/practice_iterations/two_pointers_5/TripletsWithSmallerSum.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(nlogn) / O(1)
public class TripletsWithSmallerSum {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int s = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        Arrays.sort(a);

        int cnt = 0;
        for (int i = 0; i < n - 2; i++) {
          int left = i + 1, right = n - 1;

          while (left < right) {
            int sum = a[i] + a[left] + a[right];

            if (sum < s) {
              cnt += right - left;
              left++;
            } else {
              right--;
            }
          }
        }

        System.out.println(cnt);
      }
    }
  }
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_1/triplets_with_smaller_sum.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

// O(n2) / O(1)
int count_less_than_target(vector<int> &a, int s) {
  sort(a.begin(), a.end());

  int n = a.size(), cnt = 0;
  for (int i = 0; i < n - 2; i++) {
    int left = i + 1, right = n - 1;

    while (left < right) {
      int sum = a[i] + a[left] + a[right];
      if (sum < s) {
        cnt += right - left;
        left++;
      } else {
        right--;
      }
    }
  }

  return cnt;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, s;
  cin >> tt;

  while (tt--) {
    cin >> n >> s;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    cout << count_less_than_target(a, s) << "\n";
  }

  return 0;
}